docs/src/developers/style.md

```diff
@@ -51,6 +51,162 @@ format("test")
+correctness problems if the input type does not satisfy assumptions made by the
+author of the function.
+
+As a motivating example, consider the following function:
+```jldoctest my_sum
+julia> function my_sum(x)
+           y = 0.0
+           for i in 1:length(x)
+               y += x[i]
+           end
+           return y
+       end
+my_sum (generic function with 1 method)
+```
+This function contains a number of implicit assumptions about the type of `x`:
+ * `x` supports 1-based indexing and implements `length`
+ * The element type of `x` supports addition with `0.0`, and then with the
+   result of `x + 0.0`.
+
+(As a motivating example for the second point, [`VariableRef`](@ref) plus
+`Float64` produces an [`AffExpr`](@ref). Do not assume that `+(::T, ::T)`
+produces the type `T`.)
+
+`my_sum` works as expected if the user passes in `Vector{Float64}`:
+```jldoctest my_sum
+julia> my_sum([1.0, 2.0, 3.0])
+6.0
+```
+but it doesn't respect input types, for example returning a `Float64` if the
+user passes `Vector{Int}`:
+```jldoctest my_sum
+julia> my_sum([1, 2, 3])
+6.0
+```
+and it violdates the [MethodError principle](@ref method_error_principle):
+```jldoctest my_sum
+julia> my_sum("abc")
+ERROR: MethodError: no method matching +(::Float64, ::Char)
+[...]
+```
+
+The JuMP style guide has two main suggestions for writing and interacting with
+generic code.
+
+**Option 1: Program restrictively and let users extend new methods.**
+
+In this scenario, we explictly restrict input arguments to types that are tested
+and have been validated for correctness. For example:
+```jldoctest my_sum
+julia> function my_sum_restrictive(x::Vector{Float64})
+           y = 0.0
+           for i in 1:length(x)
+               y += x[i]
+           end
+           return y
+       end
+my_sum_restrictive (generic function with 1 method)
+
+julia> my_sum_restrictive([1.0, 2.0, 3.0])
+6.0
+```
+
+Programming restrictively satisfies the
+[MethodError principle](@ref method_error_principle):
+```jldoctest my_sum
+julia> my_sum_restrictive("abc")
+ERROR: MethodError: no method matching my_sum_restrictive(::String)
+```
+and it allows other types to be supported in future by defining new methods:
+```jldoctest my_sum
+julia> function my_sum_restrictive(x::Array{T,N}) where {T,N}
+           y = zero(T)
+           for i in 1:length(x)
+               y += x[i]
+           end
+           return y
+       end
+my_sum_restrictive (generic function with 2 methods)
+```
+Importantly, these methods do not have to be defined in the original package.
+
+**Option 2: Program defensively, and validate all assumptions.**
+
+An alternative is to program defensively, and to rigorously list and validate
+all assumptions that the code makes. For example:
+```jldoctest my_sum
+julia> function my_sum_defensive(x::AbstractArray{T}) where {T}
+           try
+               # Some types may not define zero.
+               @assert zero(T) isa T
```

Comment:
I'd reorder the priorities: 1) All assumptions on abstract types (that aren't guaranteed by the definition of the abstract type) should be documented. 2) If practical, also check the assumptions in code. 3) It's good practice to add test coverage for a range of corner cases.